drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +47,617 @@ extern "C" {
+                    let ram_end_offset = ram.len().min(end_offset);
+                    // Transfer starts within the RAM part of this image.
+                    let data = ram
+                        .get((start_offset)..ram_end_offset)
+                        .ok_or(UpdateError::OutOfBounds)?;
+                    buffer.copy_from_slice(data);
+                    remainder = remainder
+                        .checked_sub(data.len())
+                        .ok_or(UpdateError::OutOfBounds)?;
+                    start_offset = ram_end_offset;
+                }
+                // Transfer data from the flash-backed portion of the image.
+                if remainder > 0 {
+                    let start = span
+                        .stored
+                        .start
+                        .checked_add(start_offset as u32)
+                        .ok_or(UpdateError::OutOfBounds)?;
+                    let end = start
+                        .checked_add(remainder as u32)
+                        .ok_or(UpdateError::OutOfBounds)?;
+                    if span.stored.contains(&start)
+                        && (span.stored.start..=span.stored.end).contains(&end)
+                    {
+                        indirect_flash_read(flash, start, buffer)?;
+                    } else {
+                        return Err(UpdateError::OutOfBounds);
+                    }
+                }
+                Ok(())
+            }
+        }
+    }
+
+    /// Get the rounded up length of an LPC55 image if present.
+    pub fn padded_image_len(&self) -> Result<u32, UpdateError> {
+        let vectors = match self.accessor {
+            Accessor::Flash { .. } => {
+                let buffer =
+                    &mut [0u8; core::mem::size_of::<ImageVectorsLpc55>()];
+                self.read_bytes(0u32, buffer.as_bytes_mut())?;
+                ImageVectorsLpc55::read_from_prefix(&buffer[..])
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+            Accessor::Ram { buffer, .. } | Accessor::Hybrid { buffer, .. } => {
+                ImageVectorsLpc55::read_from_prefix(buffer)
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+        }?;
+        let len = vectors.image_length().ok_or(UpdateError::BadLength)?;
+        round_up_to_flash_page(len).ok_or(UpdateError::BadLength)
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC value format.
+//
+// Invalid EPOC values converted to Epoch{value:0} include:
+//   - empty slice
+//   - any non-ASCII-digits in slice
+//   - any non-UTF8 in slice
+//   - leading '+' normally allowed by parse()
+//   - values greater than u32::MAX
+//
+//   Hand coding reduces size by about 950 bytes.
+impl From<&[u8]> for Epoch {
```

Comment:
I would vote to return an `Option<Epoch>` instead, or `Result<Epoch, ()>` if you want to use `TryFrom`.  Snapping invalid epoch tags to 0 seems like a recipe for confusion, because it will allow updates to an invalid epoch if the initial epoch is 0.